File: nasdaq-taq/src/main/java/org/jvirtanen/nasdaq/taq/TAQSink.java
package org.jvirtanen.nasdaq.taq;

import com.paritytrading.parity.file.taq.TAQ;
import com.paritytrading.parity.file.taq.TAQWriter;
import com.paritytrading.parity.top.MarketListener;
import com.paritytrading.parity.top.Side;
import java.io.Flushable;

class TAQSink implements Flushable, MarketListener {

    private static final long NANOS_PER_MILLI = 1000 * 1000;

    private TAQ.Quote quote;

    private TAQ.Trade trade;

    private TAQWriter writer;

    private long timestampHigh;
    private long timestampLow;

    TAQSink(String date, String instrument) {
        quote = new TAQ.Quote();
        trade = new TAQ.Trade();

        quote.date = date;
        trade.date = date;

        quote.instrument = instrument;
        trade.instrument = instrument;

        writer = new TAQWriter(System.out);
    }

    public void timestamp(long high, long low) {
        timestampHigh = high;
        timestampLow  = low;
    }

    @Override
    public void bbo(long instrument, long bidPrice, long bidSize, long askPrice, long askSize) {
        quote.timestampMillis = timestampMillis();

        quote.bidPrice = bidPrice;
        quote.bidSize  = bidSize;
        quote.askPrice = askPrice;
        quote.askSize  = askSize;

        writer.write(quote);
    }

    @Override
    public void trade(long instrument, Side side, long price, long size) {
        trade.timestampMillis = timestampMillis();

        trade.price = price;
        trade.size  = size;
        trade.side  = side(side);

        writer.write(trade);
    }

    public void trade(long price, long size) {
        trade.timestampMillis = timestampMillis();

        trade.price = price;
        trade.size  = size;
        trade.side  = TAQ.UNKNOWN;

        writer.write(trade);
    }

    @Override
    public void flush() {
        writer.flush();
    }

    private long timestampMillis() {
        return ((timestampHigh << 32) + timestampLow) / NANOS_PER_MILLI;
    }

    private char side(Side side) {
        switch (side) {
        case BUY:
            return TAQ.BUY;
        case SELL:
            return TAQ.SELL;
        }

        return TAQ.UNKNOWN;
    }

}

File: nasdaq-taq/src/main/kotlin/org/jvirtanen/nasdaq/taq/TAQSink.kt
package org.jvirtanen.nasdaq.taq

import com.paritytrading.parity.file.taq.TAQ
import com.paritytrading.parity.file.taq.TAQWriter
import com.paritytrading.parity.top.MarketListener
import com.paritytrading.parity.top.Side
import java.io.Flushable

val NANOS_PER_MILLI = 1000 * 1000

class TAQSink(date: String, instrument: String): Flushable, MarketListener {

    private val quote = TAQ.Quote()
    private val trade = TAQ.Trade()

    private val writer = TAQWriter(System.out)

    private var timestampHigh = 0L
    private var timestampLow  = 0L

    init {
        quote.date = date
        trade.date = date

        quote.instrument = instrument
        trade.instrument = instrument
    }

    fun timestamp(high: Int, low: Long) {
        timestampHigh = high.toLong()
        timestampLow  = low
    }

    override fun bbo(instrument: Long, bidPrice: Long, bidSize: Long, askPrice: Long, askSize: Long) {
        quote.timestampMillis = timestampMillis()

        quote.bidPrice = bidPrice
        quote.bidSize  = bidSize
        quote.askPrice = askPrice
        quote.askSize  = askSize

        writer.write(quote)
    }

    override fun trade(instrument: Long, side: Side, price: Long, size: Long) {
        trade.timestampMillis = timestampMillis()

        trade.price = price
        trade.size  = size
        trade.side  = side(side)

        writer.write(trade)
    }

    fun trade(price: Long, size: Long) {
        trade.timestampMillis = timestampMillis()

        trade.price = price
        trade.size  = size
        trade.side  = TAQ.UNKNOWN

        writer.write(trade)
    }

    override fun flush() {
        writer.flush()
    }

    private fun timestampMillis(): Long {
        return ((timestampHigh shl 32) + timestampLow) / NANOS_PER_MILLI
    }

    private fun side(side: Side): Char {
        when (side) {
            Side.BUY  -> return TAQ.BUY
            Side.SELL -> return TAQ.SELL
            else      -> return TAQ.UNKNOWN
        }
    }

}
